enforce batch header fields

tbg opened the pull request:
- User, UserPrio and Txn cause errors when sent through the TxnCoordSender
  in an individual call which is part of a Batch with non-identical header
- Had to leave them in the individual calls in store.resolveWriteIntent because of
  tests in store_test.go which omit the TxnCoordSender; added a TODO there.
  Batch handling and/or those tests need some additional cleanup beyond this PR.

bdarnell commented on kv/txn_coord_sender.go:
```diff
@@ -407,20 +405,26 @@ func (tc *TxnCoordSender) sendBatch(batchArgs *proto.InternalBatchRequest, batch
-		if args.Header().Txn == nil {
-			args.Header().Txn = batchArgs.Txn
+		args.Header().UserPriority = batchArgs.UserPriority
+		if args.Header().Txn != nil && args.Header().GetTxn() != batchArgs.GetTxn() {
```
Does `!=` do the right thing with the Transaction struct? I think it's a shallow compare which would treat two equivalent transactions as different because the pointers in their byte slices are different.

tschottdorf replied:
That's why I used GetTxn, thinking it would return a `proto.Transaction`. I'll update accordingly.

bdarnell replied:
Even if it returned a `proto.Transaction` instead of a `*proto.Transaction`, there would be a problem, because the Transaction struct includes fields with pointer types. These will compare as equal only if they point to the same memory (which may be true in local-only tests, but will not be true once things have passed through the DistSender and been marshaled and unmarshaled).

tschottdorf replied:
That's okay though, usually we compare only the Transaction ID.

bdarnell commented:
LGTM

tbg committed:
enforce batch header fields

- User, UserPrio and Txn cause errors when sent through the TxnCoordSender
  in an individual call which is part of a Batch with non-identical header
- Had to leave double-specify them in store.resolveWriteIntent because of
  tests in store_test.go which omit the TxnCoordSender; added a TODO there.

tbg committed:
add test for Transaction.Equal
